backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java

```diff
@@ -1967,97 +1970,138 @@ public void onClick(DialogInterface dialog, int which) {
 		return true;
 	}
 
-
-	private boolean containsStringEntry(@NonNull String[] stringItenary, String targetEntry) {
-		for (String sourceEntry : stringItenary) {
-			// Log.d(ScummVM.LOG_TAG, "Comparing filename: " + sourceEntry + " to filename: " + targetEntry);
-			if (sourceEntry.compareToIgnoreCase(targetEntry) == 0) {
-				return true;
+	private static void deleteDir(File dir) {
+		for (File child : dir.listFiles()) {
+			if (child.isDirectory()) {
+				deleteDir(child);
+			} else {
+				if (!child.delete()) {
+					Log.e(ScummVM.LOG_TAG, "Failed to delete file:" + child.getPath());
+				}
 			}
 		}
-		return false;
+		if (!dir.delete()) {
+			Log.e(ScummVM.LOG_TAG, "Failed to delete dir:" + dir.getPath());
+		}
 	}
 
 	// clear up any possibly deprecated assets (when upgrading to a new version)
 	// Don't remove the scummvm.ini nor the scummvm.log file!
 	// Remove any files not in the filesItenary, even in a sideUpgrade
 	// Remove any files in the filesItenary only if not a sideUpgrade
-	private void internalAppFolderCleanup(String[] filesItenary, boolean sideUpgrade) {
-		if (_actualScummVMDataDir != null) {
-			File[] extfiles = _actualScummVMDataDir.listFiles();
-			if (extfiles != null) {
-				Log.d(ScummVM.LOG_TAG, "Cleaning up files in internal app space");
-				for (File extfile : extfiles) {
-					if (extfile.isFile()) {
-						if (extfile.getName().compareToIgnoreCase("scummvm.ini") != 0
-							&& extfile.getName().compareToIgnoreCase("scummvm.log") != 0
-							&& (!containsStringEntry(filesItenary, extfile.getName())
-							|| !sideUpgrade)
-						) {
-							Log.d(ScummVM.LOG_TAG, "Deleting file:" + extfile.getName());
-							if (!extfile.delete()) {
-								Log.e(ScummVM.LOG_TAG, "Failed to delete file:" + extfile.getName());
-							}
-						}
-					}
+	private static void internalAppFolderCleanup(boolean sideUpgrade, File dataDir, String[] filesItenary) {
+		HashSet<String> filesToKeep = new HashSet<>(Arrays.asList(filesItenary));
```

Comment:
Should we also handle here the case when filesItenary is null (because eg. the IOException was thrown in extractAssets() or if the assetManager.list(assetDir) call returns null)? 
I think in this case, we still want the cleanup, and the filesToKeep (which refers only to keeping files in the case of a side-upgrade) will be an empty array.

Reply:
I slightly changed the working here.
I call this function only when `filesItenary`  is non null and not empty.
If it's null, it means that the asset is a file: in this case I delete the directory matching the name.
If it is already a file, it will get overwritten.